src/dmd/dimport.d

```diff
@@ -237,8 +237,41 @@ extern (C++) final class Import : Dsymbol
             protection = sc.protection;
         if (!isstatic && !aliasId && !names.dim)
             sc.scopesym.importScope(mod, protection);
+        // Enable access to pkgs/mod as soon as posible, because compiler
+        // can traverse them before the import gets semantic (Issue: 21501)
+        if (!aliasId && !names.dim)
+            addPackageAccess(sc.scopesym);
     }
 
+    /*******************************
+     * Mark the imported packages as accessible from the current
+     * scope. This access check is necessary when using FQN b/c
+     * we're using a single global package tree.
+     * https://issues.dlang.org/show_bug.cgi?id=313
+     */
+    void addPackageAccess(ScopeDsymbol scopesym)
```

Comment:
This could probably be private (this would also solve the need to update the headers).

Reply:
I made it `extern (D)`.

Reply:
That also works.